{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1c648442-0f2a-44d1-9fe8-40a642291e0e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/016651544/anaconda3/envs/llama2/lib/python3.10/site-packages/transformers/utils/generic.py:260: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.\n",
      "  torch.utils._pytree._register_pytree_node(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEVICE:-  cuda:0\n"
     ]
    }
   ],
   "source": [
    "from llama2_fine_tuning_main import pick_balanced_data,read_data, balanced_split, create_model_and_tokenizer\n",
    "import json\n",
    "import re\n",
    "from pprint import pprint\n",
    "\n",
    "import pandas as pd\n",
    "import torch\n",
    "from datasets import Dataset, load_dataset\n",
    "from huggingface_hub import notebook_login\n",
    "from peft import LoraConfig, PeftModel, AutoPeftModelForCausalLM\n",
    "from transformers import (\n",
    "    AutoModelForCausalLM,\n",
    "    AutoTokenizer,\n",
    "    BitsAndBytesConfig,\n",
    "    TrainingArguments,\n",
    ")\n",
    "\n",
    "from trl import SFTTrainer\n",
    "import csv\n",
    "from datasets import Dataset, DatasetDict\n",
    "import random\n",
    "import locale\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "de1bf596-d357-4f52-ad48-4bff7e2a51a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEVICE:-  cuda:0\n"
     ]
    }
   ],
   "source": [
    "# login(\"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\")\n",
    "DEVICE = \"cuda:0\" if torch.cuda.is_available() else \"cpu\"\n",
    "print(\"DEVICE:- \",DEVICE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "44a4ac54-7ccf-40f8-afb4-7c6fc2a202c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "!export http_proxy=http://172.16.1.2:3128; export https_proxy=http://172.16.1.2:3128\n",
    "!export HUGGINGFACE_HUB_TOKEN=\"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\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3552115f-f301-4728-aaaa-e991b62cc573",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from huggingface_hub import HfApi, HfFolder\n",
    "\n",
    "# HfFolder.set_token(\"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\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "2c711975-6266-4135-8c58-d47dfbb8a2fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "DEFAULT_SYSTEM_PROMPT = \"\"\"\n",
    "Below is a sentence with biomedical entities masked as $GENE$ and $DISEASE$ with biomedical related text. Return boolean variable True/False if there is a relation between these two entities (masked variables) based on context. .\n",
    "\"\"\".strip()\n",
    "\n",
    "\n",
    "def generate_test_prompt(\n",
    "    conversation: str, system_prompt: str = DEFAULT_SYSTEM_PROMPT\n",
    ") -> str:\n",
    "    return f\"\"\"### Instruction: {system_prompt}\n",
    "\n",
    "### Input:\n",
    "{conversation.strip()}\n",
    "\n",
    "### Response:\n",
    "\"\"\".strip()\n",
    "\n",
    "def generate_test_feature(data):\n",
    "    return [{'sentence':item['sentence'],'response':item['response'],'text': generate_test_prompt(item['sentence'],item['instruction'])} for item in data]\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ce3cc95a-be2f-4c42-9333-b395975f5848",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "TRAIN_SAMPLES=250\n",
    "VAL_SAMPLES=50\n",
    "TEST_SAMPLES=1000\n",
    "EPOCHS=25\n",
    "BATCH_SIZE=16\n",
    "TRAIN_FILE_PATH=\"./train_data.tsv\" \n",
    "TEST_FILE_PATH=\"./test_data.tsv\"\n",
    "# OUTPUT_DIR=\"./experiments/1000_25_FT_model/checkpoint-200/\"\n",
    "OUTPUT_DIR=\"./experiments/10000_25_FT_model/\"\n",
    "MODEL_NAME=\"meta-llama/Llama-2-7b-hf\"\n",
    "SAVE_FILEPATH=\"./test_df_10000_25_FT_model.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "549b1a82-8724-412d-bbda-f12198801ee6",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_samples = 250\n",
    "val_samples = 50\n",
    "test_samples = 2660 #5330\n",
    "n_epochs = 25#args.n_epochs\n",
    "batch_size = 16#args.batch_size\n",
    "\n",
    "# Replace 'your_file.tsv' with the path to your actual file\n",
    "# train_file_path = args.train_file_path\n",
    "test_file_path = \"./test_data.tsv\" #args.test_file_path\n",
    "OUTPUT_DIR = \"./experiments/10000_25_FT_model/\" #args.output_dir #f\"./experiments/{train_samples}_{n_epochs}_FT_model\"\n",
    "MODEL_NAME = \"meta-llama/Llama-2-7b-hf\" #args.model_name#\"meta-llama/Llama-2-7b-hf\"\n",
    "save_filepath = \"./test_df_10000_25_FT_model.csv\"#args.save_filepath #f\"./test_df_{train_samples}_{n_epochs}_FT_model.csv\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "8cfbe07b-0b9f-4fc4-8f3b-5d054343d75b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5330\n"
     ]
    }
   ],
   "source": [
    "test_data = read_data(test_file_path)\n",
    "print(len(test_data))\n",
    "# balanced_data = pick_balanced_data(test_data, count=test_samples)\n",
    "# test_dataset = generate_test_feature(balanced_data)\n",
    "test_dataset = generate_test_feature(test_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "c1713673-fec3-4997-af77-c8a65bc7aa92",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e897e9af872f4b18998226b3c6818e21",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### Instruction: You will be given a sentence. Your job is to return a boolean variable True/False indicating if there is a relation between the two masked variables.\n",
      "\n",
      "### Input:\n",
      "These results suggest that the C1772T polymorphism in @GENE$ is not involved in progression or metastasis of @DISEASE$.\n",
      "\n",
      "### Response:\n"
     ]
    }
   ],
   "source": [
    "\n",
    "model, tokenizer = create_model_and_tokenizer(MODEL_NAME)\n",
    "model.config.use_cache = False\n",
    "\n",
    "model = PeftModel.from_pretrained(model, OUTPUT_DIR)\n",
    "print(test_dataset[0]['text'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "11696c01-d1af-4cd5-9253-779d3c72e419",
   "metadata": {},
   "outputs": [],
   "source": [
    "def summarize(model, text: str):\n",
    "    inputs = tokenizer(text, return_tensors=\"pt\").to(DEVICE)\n",
    "    inputs_length = len(inputs[\"input_ids\"][0])\n",
    "    \n",
    "    with torch.inference_mode():\n",
    "        start_time = time.time()\n",
    "        outputs = model.generate(**inputs, max_new_tokens=64, temperature=0.0001)\n",
    "        end_time = time.time()\n",
    "        generate_time = end_time - start_time\n",
    "    # print(f\"Tokenizer time: {tokenizer_time:.2f} seconds\")\n",
    "    print(f\"Generation time: {generate_time:.2f} seconds\")\n",
    "    # print(outputs)\n",
    "    return tokenizer.decode(outputs[0][inputs_length:], skip_special_tokens=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "322fe3e6-c7ba-409a-bc70-d69be17752fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generation time: 7.13 seconds\n",
      "Time taken to summarize item 1: 7.13 seconds\n",
      "{'sentence': 'These results suggest that the C1772T polymorphism in @GENE$ is not involved in progression or metastasis of @DISEASE$.', 'actual_output': 'True', 'predicted_output': 'True'}\n",
      "Generation time: 7.13 seconds\n",
      "Time taken to summarize item 2: 7.14 seconds\n",
      "{'sentence': 'In our setting, @DISEASE$ among alcoholic individuals seems to be independent of the presence of mutations C282Y, H63D and S65C in the @GENE$ gene.', 'actual_output': 'True', 'predicted_output': 'True'}\n",
      "Generation time: 7.14 seconds\n",
      "Time taken to summarize item 3: 7.14 seconds\n",
      "{'sentence': 'MPO genotype GG is associated with @DISEASE$ in patients with hereditary @GENE$.', 'actual_output': 'True', 'predicted_output': 'True'}\n",
      "Generation time: 7.19 seconds\n",
      "Time taken to summarize item 4: 7.19 seconds\n",
      "{'sentence': 'These three studies do not provide consistent evidence supporting the hypothesis that @GENE$ mutations are associated with an increased risk of @DISEASE$ and with the development of arteriosclerosis.', 'actual_output': 'True', 'predicted_output': 'True'}\n",
      "Generation time: 7.13 seconds\n",
      "Time taken to summarize item 5: 7.13 seconds\n",
      "{'sentence': 'Our prospective findings suggest that individuals carrying the @GENE$ C282Y mutation may be at increased risk of @DISEASE$.', 'actual_output': 'True', 'predicted_output': 'True'}\n",
      "Generation time: 7.23 seconds\n",
      "Time taken to summarize item 6: 7.23 seconds\n",
      "{'sentence': 'We conclude that  homozygosity for the G1514-->A mutation is exclusively responsible for the adult form of @DISEASE$ in this family, and that the A619-->G substitution is not a deleterious mutation but rather a common @GENE$ polymorphism.', 'actual_output': 'True', 'predicted_output': 'True'}\n",
      "Generation time: 7.14 seconds\n",
      "Time taken to summarize item 7: 7.15 seconds\n",
      "{'sentence': 'The data suggest that the @GENE$ gene or a linked locus significantly modulates the risk for @DISEASE$.', 'actual_output': 'True', 'predicted_output': 'True'}\n",
      "Generation time: 7.14 seconds\n",
      "Time taken to summarize item 8: 7.14 seconds\n",
      "{'sentence': 'The novel gene @GENE$ may be related with the infiltration and proliferation of @DISEASE$.', 'actual_output': 'True', 'predicted_output': 'True'}\n",
      "Generation time: 7.15 seconds\n",
      "Time taken to summarize item 9: 7.15 seconds\n",
      "{'sentence': 'Our findings suggest that the genetic variants of the @GENE$ but not the TIM-3 gene contribute to @DISEASE$ susceptibility in this African-American population.', 'actual_output': 'True', 'predicted_output': 'True'}\n",
      "Generation time: 7.14 seconds\n",
      "Time taken to summarize item 10: 7.14 seconds\n",
      "{'sentence': 'In conclusion, the M416V polymorphism of @GENE$ gene is not associated with insulin resistance in @DISEASE$.', 'actual_output': 'True', 'predicted_output': 'True'}\n",
      "Generation time: 7.14 seconds\n",
      "Time taken to summarize item 11: 7.14 seconds\n",
      "{'sentence': 'We found no evidence that mutation in @GENE$,GNGT1,or RGS9 gene is a cause of @DISEASE$.', 'actual_output': 'True', 'predicted_output': 'True'}\n",
      "Generation time: 7.14 seconds\n",
      "Time taken to summarize item 12: 7.14 seconds\n",
      "{'sentence': 'These results suggest that the @GENE$ null genotype may be associated with increased risk of @DISEASE$.', 'actual_output': 'True', 'predicted_output': 'True'}\n",
      "Generation time: 7.16 seconds\n",
      "Time taken to summarize item 13: 7.16 seconds\n",
      "{'sentence': 'These findings suggest that the GSTA1 and @GENE$ polymorphisms are associated with @DISEASE$ susceptibility, especially among smokers.', 'actual_output': 'True', 'predicted_output': 'True'}\n",
      "Generation time: 7.14 seconds\n",
      "Time taken to summarize item 14: 7.14 seconds\n",
      "{'sentence': 'These findings suggest that individual susceptibility to @DISEASE$ may be modulated by GSTM1, @GENE$ and NAT2 polymorphisms.', 'actual_output': 'True', 'predicted_output': 'True'}\n",
      "Generation time: 7.16 seconds\n",
      "Time taken to summarize item 15: 7.16 seconds\n",
      "{'sentence': 'These results suggest that the @GENE$ null-genotype is associated with an increased risk of @DISEASE$, especially in younger individuals.', 'actual_output': 'True', 'predicted_output': 'True'}\n",
      "Generation time: 7.20 seconds\n",
      "Time taken to summarize item 16: 7.21 seconds\n",
      "{'sentence': 'These results support previous reports that the GSTM1 null genotype is associated with a modest increase in risk for @DISEASE$, particularly among heavy smokers, suggest no role for @GENE$ and the need for further study of GSTP1.', 'actual_output': 'True', 'predicted_output': 'True'}\n",
      "Generation time: 7.20 seconds\n",
      "Time taken to summarize item 17: 7.20 seconds\n",
      "{'sentence': 'We conclude that  the two polymorphisms, @GENE$ null and PON1 BB, are common genetic traits that pose low individual risk but may be important determinants of overall population @DISEASE$ risk, particularly among groups exposed to NHL-related carcinogens.', 'actual_output': 'True', 'predicted_output': 'True'}\n",
      "Generation time: 7.21 seconds\n",
      "Time taken to summarize item 18: 7.21 seconds\n",
      "{'sentence': 'We conclude that  the two polymorphisms, @GENE$ null and PON1 BB, are common genetic traits that pose low individual risk but may be important determinants of overall population NHL risk, particularly among groups exposed to @DISEASE$-related carcinogens.', 'actual_output': 'True', 'predicted_output': 'True'}\n",
      "Generation time: 7.20 seconds\n",
      "Time taken to summarize item 19: 7.20 seconds\n",
      "{'sentence': 'Polymorphisms in the oxidative stress-related genes (CYP1A1, GSTM1, @GENE$, MPO, MnSOD) do not seem to be risk factors for @DISEASE$.', 'actual_output': 'True', 'predicted_output': 'True'}\n",
      "Generation time: 7.22 seconds\n",
      "Time taken to summarize item 20: 7.22 seconds\n",
      "{'sentence': 'The @GENE$-null genotype may increase the risk for @DISEASE$ and is associated with favorable prognostic factors, and the presence of at least one GST deletion indicates an improved disease-free survival.', 'actual_output': 'True', 'predicted_output': 'True'}\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "summaries = []\n",
    "new_rows = []\n",
    "test_df = pd.DataFrame(columns = ['sentence','actual_output','predicted_output'])\n",
    "for i,item in enumerate(test_dataset):\n",
    "    start_time = time.time()  # Record the start time\n",
    "    summary = summarize(model, item['text'])\n",
    "    end_time = time.time()  # Record the end time\n",
    "    elapsed_time = end_time - start_time  # Calculate the elapsed time\n",
    "    print(f\"Time taken to summarize item {i + 1}: {elapsed_time:.2f} seconds\")\n",
    "    \n",
    "    summaries.append(summary.split(\"\\n\")[1])\n",
    "    new_row = {\n",
    "        'sentence': item['sentence'],\n",
    "        'actual_output': item['response'],\n",
    "        'predicted_output': summary.split(\"\\n\")[1]\n",
    "    }\n",
    "    new_rows.append(new_row)\n",
    "    print(new_row)\n",
    "new_rows_df = pd.DataFrame(new_rows)\n",
    "test_df = pd.concat([test_df, new_rows_df], ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "878f3036-c114-443f-b370-48c0610c22d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from sklearn.metrics import precision_score, recall_score, f1_score\n",
    "\n",
    "# Assuming 'df' is your DataFrame\n",
    "actual = test_df['actual_output']#.astype(int)\n",
    "predicted = test_df['predicted_output'].str.replace(r\"[^TrueFalse]\", \"\", regex=True)#.astype(int)\n",
    "precision = precision_score(actual, predicted,average='weighted')\n",
    "recall = recall_score(actual, predicted,average='weighted')\n",
    "f1 = f1_score(actual, predicted,average='weighted')\n",
    "\n",
    "print(f\"Precision: {precision}\")\n",
    "print(f\"Recall: {recall}\")\n",
    "print(f\"F1 Score: {f1}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c960979f-73e9-4bff-9217-88f34bbe4e8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test_df.to_csv(save_filepath, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "llama2",
   "language": "python",
   "name": "llama2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
